### DISCUSSION TITLE
API documentation behind Nginx reverse proxy

### DISCUSSION THREAD

**Discussion by christophelebrun** [fastapi/fastapi#8593] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

#### Python file (last_signal_api.py)

```python
from typing import List
from fastapi import FastAPI, Query

app = FastAPI() 

@app.get("/")
async def get_last_signal(mmsi: List[int] = Query(None)) :
    return {}
```

#### Nginx (reverse proxy) file (/etc/nginx/sites-enabled/default):
```
location /last_signal/docs {
        proxy_pass http://127.0.0.1:8002/docs;
}

location /last_signal {
        proxy_pass http://127.0.0.1:8002/?$args;
}
```

#### Uvicorn command:
```
nohup uvicorn last_signal_api:app --port 8002 --root-path="/last_signal"
```

### Description

Hello,

I have troubles to reach the openapi.json file for documentation when my API is running behind a nginx reverse-proxy.
I have read several pages (including https://fastapi.tiangolo.com/advanced/behind-a-proxy/), but I am still not able to get it right.

My API works fine, but I got a "Failed to load API definition. Fetch error Not Found /openapi.json" errror when trying to reach the endpoint for documentation (in my case: http://heg-rl001.hesge.ch/last_signal/docs).

I tried:
* to set the --root-path option for uvicorn
* to set the --root-path option for FastAPI()
* set none or both

But it didn't solved the problem.

All advice will be welcomed !

### Operating System

Linux

### Operating System Details

Ubuntu

### FastAPI Version

FastAPI 0.73.0

### Python Version

Python 3.8.10

**Comment by dajtxx** [comment]
I've hit something similar. I'm not happy with my solution, but here it is.

In my py module I set things so the API is at https://host/broker/api

```
router = APIRouter(prefix='/broker/api')

@router.get('/something/')
async def get_something() {
    ...
}

app = FastAPI(title='My API', version='1.0.0')
app.include_router(router)
```

In nginx I have numerous locations to work with this. 

```
    location /broker/api/docs {
        proxy_pass http://localhost:9999/docs;
    }

    location /openapi.json {
        proxy_pass http://localhost:9999/openapi.json;
    }

    location /broker/ {
        proxy_pass http://localhost:9999/broker/;
    }
```

I guess the docs can't be made available at what we intuitively think of as the API root path because FastAPI is so flexible in how you set up your routes or something.

I had root_path working in a simpler config but struggled with this one.

**Comment by raminmardani** [comment]
this solution doesn't work!

**Comment by JarroVGIT** [comment]
Can you see what URL the browser tries to load when going to the docs endpoint? The HTML that is rendered wil try that URL and that might give some indication where it goes wrong. I would expect it to be something with root_path indeed, so try different combinations again and see what happens to the URL that is being trying to load.

**Comment by raminmardani** [comment]
@JarroVGIT here is my question: 
https://stackoverflow.com/questions/72734893/api-documentation-on-nginx-server

**Comment by christophelebrun** [comment]
Hello,

I found a solution (inspired from the answer of @dajtxx) where I add a clause for the /openapi.json redirection:

**Nginx (reverse proxy) file (/etc/nginx/sites-enabled/default):**

```
location /last_signal/docs {
        proxy_pass http://127.0.0.1:8002/docs;
}

location /last_signal/openapi.json {
                proxy_pass http://127.0.0.1:8001/openapi.json;
}

location /last_signal {
        proxy_pass http://127.0.0.1:8002/?$args;
}

```

I still need the --root-path argument when starting the API:
`nohup uvicorn last_signal_api:app --port 8002 --root-path="/last_signal"`

**Comment by cbadenes** [comment]
Hi, 

This is my solution, slightly different from the previous one, in case it would be useful. 

The content of the `/etc/nginx/sites-enabled/default`  file is: 

```yml

       location /last_signal/ {
		rewrite ^/last_signal(.*) $1 break;
		proxy_pass http://127.0.0.1:8000;
	}
       
	location /last_signal {
		rewrite ^/last_signal(.*) /$1 break;
		proxy_pass http://127.0.0.1:8000;
	}
```

and the --root-path argument is /last_signal:
` nohup uvicorn last_signal_api:app --port 8002 --root-path="/last_signal"`

**Comment by vedil** [comment]
adding the below code
```
@app.get("/docs", include_in_schema=False)
async def custom_swagger_ui_html():
    logger.debug("app.openapi_url is %s", app.openapi_url)
    return get_swagger_ui_html(
        openapi_url=app.openapi_url,
        title=app.title + " - Swagger UI",
        oauth2_redirect_url=app.swagger_ui_oauth2_redirect_url,
        swagger_js_url="https://unpkg.com/swagger-ui-dist@5/swagger-ui-bundle.js",
        swagger_css_url="https://unpkg.com/swagger-ui-dist@5/swagger-ui.css",
    )


@app.get(app.swagger_ui_oauth2_redirect_url, include_in_schema=False)
async def swagger_ui_redirect():
    return get_swagger_ui_oauth2_redirect_html()


@app.get("/redoc", include_in_schema=False)
async def redoc_html():
    return get_redoc_html(
        openapi_url=app.openapi_url,
        title=app.title + " - ReDoc",
        redoc_js_url="https://unpkg.com/redoc@next/bundles/redoc.standalone.js",
    )
```

this will help https://github.com/tiangolo/fastapi/issues/4211

also i am caching the instance of openapi  where i append servers 
like

`
openapi_schema = get_openapi(title="blah", routes=app.routes,
            servers=app.servers + [{"url": "/"}],
`

that worked for me

**Comment by YuriiMotov** [comment, marked as answer]
The way FastAPI handles `root_path` has been updated since that time. So, I believe this is outdated.
If you still struggle with this, read [docs](https://fastapi.tiangolo.com/advanced/behind-a-proxy/), search for newer discussions on topic or open new discussion
